src/Update/RecipePatcher.php

```diff
@@ -0,0 +1,220 @@
+use Symfony\Component\Filesystem\Filesystem;
+
+class RecipePatcher
+{
+    private string $rootDir;
+    private Filesystem $filesystem;
+    private IOInterface $io;
+    private ProcessExecutor $processExecutor;
+
+    public function __construct(string $rootDir, IOInterface $io)
+    {
+        $this->rootDir = $rootDir;
+        $this->filesystem = new Filesystem();
+        $this->io = $io;
+        $this->processExecutor = new ProcessExecutor($io);
+    }
+
+    /**
+     * Applies the patch. If it fails unexpectedly, an exception will be thrown.
+     *
+     * @return bool returns true if fully successful, false if conflicts were encountered
+     */
+    public function applyPatch(RecipePatch $patch): bool
+    {
+        if (!$patch->getPatch()) {
+            // nothing to do!
+            return true;
+        }
+
+        $addedBlobs = $this->addMissingBlobs($patch->getBlobs());
+
+        $patchPath = $this->rootDir.'/_flex_recipe_update.patch';
+        file_put_contents($patchPath, $patch->getPatch());
+
+        try {
+            $this->execute('git update-index --refresh', $this->rootDir);
+
+            $output = '';
+            $statusCode = $this->processExecutor->execute('git apply "_flex_recipe_update.patch" -3', $output, $this->rootDir);
+            $this->io->debug('git apply output: '.$output);
+
+            if (0 === $statusCode) {
+                // successful with no conflicts
+                return true;
+            }
+
+            if (str_contains($this->processExecutor->getErrorOutput(), 'with conflicts')) {
+                // successful with conflicts
+                return false;
+            }
+
+            throw new \LogicException('Error applying the patch: '.$this->processExecutor->getErrorOutput());
+        } finally {
+            unlink($patchPath);
+            // clean up any temporary blobs
+            foreach ($addedBlobs as $filename) {
+                unlink($filename);
+            }
+        }
+    }
+
+    public function generatePatch(array $originalFiles, array $newFiles): RecipePatch
+    {
+        // null implies "file does not exist"
+        $originalFiles = array_filter($originalFiles, function ($file) {
+            return null !== $file;
+        });
+        $newFiles = array_filter($newFiles, function ($file) {
+            return null !== $file;
+        });
+
+        // find removed files and add them so they will be deleted
+        foreach ($originalFiles as $file => $contents) {
+            if (!isset($newFiles[$file])) {
+                $newFiles[$file] = null;
+            }
+        }
+
+        // If a file is being modified, but does not exist in the current project,
+        // it cannot be patched. Ignore these files and leave them deleted.
+        $modifiedFiles = array_intersect_key(array_keys($originalFiles), array_keys($newFiles));
+        foreach ($modifiedFiles as $modifiedFile) {
+            if (!file_exists($this->rootDir.'/'.$modifiedFile)) {
```

Comment:
Would be great to not lose the patch when the file does not exist!

My used case: I switched from .yaml to .PHP configuration file. But I would appreciate having a diff of changes. In that way, I can implement the diff in PHP.

Logic could be something like:
if file exists in both `$originalFiles` and `$newFiles` BUT does not exists on file system (it means the USER deleted it) then dump a `.patch` somewhere (in file system, or stdout)

Reply:
Hmm, that’s an interesting idea! I cannot apply the patch - it just doesn’t work - but I could dump a patch file.

Reply:
<img width="1127" alt="Screen Shot 2021-12-18 at 4 01 46 PM" src="https://user-images.githubusercontent.com/121003/146655533-ef507de2-4169-4fe7-b105-7de86fca16bc.png">

How about this :)

Reply:
that's perfect ❤️